python/lib/core/dmod/core/dataset.py

```diff
@@ -487,6 +488,252 @@ def unlink_to_dataset(self, dataset: Dataset) -> bool:
         return dataset.manager is not None and dataset.manager.unlink_user(user=self, dataset=dataset)
 
 
+DataItem = TypeVar('DataItem', bound=Union[bytes, RepeatableReader, Path])
+
+
+class AbstractDomainDetector(ABC):
+    """ Abstraction for something that will automatically detect a :class:`DataDomain` for some data. """
+
+    def __init__(self, *args, **kwargs):
+        self._data_domain: Optional[DataDomain] = None
+
+    @abstractmethod
+    def detect(self, **kwargs) -> DataDomain:
+        """
+        Detect and return the data domain.
+
+        Parameters
+        ----------
+        kwargs
+            Optional kwargs applicable to the subtype, which may enhance or add to the domain detection and generation
+            capabilities, but which should not be required to produce a valid domain.
+
+        Returns
+        -------
+        DataDomain
+            The detected domain.
+
+        Raises
+        ------
+        DmodRuntimeError
+            If it was not possible to properly detect the domain.
+        """
+        pass
+
+    @property
+    def data_domain(self) -> DataDomain:
```

Comment:
I get that this is a nicety, but imo there is nothing stopping a subclass from implementing this behavior in their implementation of `detect`. This also places the burden on subtypes to update `_data_domain` when `detect` is called. I think the design can be simplified by dropping this property and the `_data_domain` class level instance variable. This could result in slightly more code, but the behavior of a subtype will be closer to its declaration.

Reply:
> I get that this is a nicety, but imo there is nothing stopping a subclass from implementing this behavior in their implementation of detect.

True.  I think my motivation is that there is nothing making them either, nor making all of them do it in a consistent manner (e.g., if you want to re-detect versus return what was detected already, can you do that, and is it done the same way?).  That seemed most easily solvable by doing some things in the superclass. 

> This also places the burden on subtypes to update `_data_domain` when `detect` is called. 

Actually the reverse is true.  Calling `detect` is intended _not_ to impact `_data_domain`.  This is why `data_domain` and `reset` are implemented:  to control (and return) the `_data_domain` value (though, as I mentioned above, I think I'm not actually using them at the moment).   But `detect` should run detection and return a new object each time.   

> I think the design can be simplified by dropping this property and the _data_domain class level instance variable. This could result in slightly more code, but the behavior of a subtype will be closer to its declaration.

Heh, we're in another thing where my style shows itself to not be especially Pythonic.  

I'd intended to leverage the superclass to abstract a concern from the subclass implementations (though if we stick with that, I do need to at least go back and use the property in places).   We could skip that, and just have the independent `detect` definitions.  Subclasses would be more free to handle themselves, but I think we'd get slightly less predictable and consistent implementations.  I'm a little worried that, given how we're using these in places like the universal detector, that inconsistency could eventually become a problem, even if it isn't right now.

Reply:
> Heh, we're in another thing where my style shows itself to not be especially Pythonic.

Haha, I don't think your style is _not_ Pythonic. I just tend to have an aversion to inheritance outside of using it strictly for enforcing an interface. Especially, when you start sharing private properties between inheritance levels. So, for sure a style thing!

> I'd intended to leverage the superclass to abstract a concern from the subclass implementations (though if we stick with that, I do need to at least go back and use the property in places). We could skip that, and just have the independent `detect` definitions. Subclasses would be more free to handle themselves, but I think we'd get slightly less predictable and consistent implementations. I'm a little worried that, given how we're using these in places like the universal detector, that inconsistency could eventually become a problem, even if it isn't right now.

First off, thanks for articulating your design so throughly! I agree that abstracting the "caching" of the detected `DataDomain` in a consistent way is valuable. I think we can achieve this behavior, in a similar fashion, without requiring the functionality be provided by the base class in a consistent way by using composition instead. This has the benefit of decoupling the "caching" behavior from the detectors and provides a separate pathway to add complexity to "caching" mechanisms if needed in the future without having to muck around with how each `AbstractDomainDetector` works. What that could look like in practice is:

```python
class CachedDomainDetector:
  def __init__(self, detector: AbstractDomainDetector):
    self._detector = detector
    self._data_domain = None

  @property
  def data_domain(self) -> DataDomain:
     if self._data_domain is None:
       self._data_domain = self._detector.detect()
     return self._data_domain

  def reset(self):
    self._data_domain = None
```

Alternatively, it could be a subclass of `AbstractDomainDetector`. So, something like:

```python
class CachedDomainDetector(AbstractDomainDetector):
  def __init__(self, detector: AbstractDomainDetector):
    self._detector = detector
    self._data_domain = None

  def detect(self, **kwargs) -> DataDomain:
     if self._data_domain is None:
       self._data_domain = self._detector.detect()
     return self._data_domain

  def reset(self):
    self._data_domain = None

```

In either case, we are able to move the base class prescribed behavior elsewhere and decouple the behavioral concepts. The trade off is now you have to know about whatever that caching wrapper type is, but I would argue that once you find it, its far easier to understand the implications of that abstraction rather than having to walk up the inheritance tree and see how each parent type handles the, in this case, `_data_domain` property.

Reply:
I went ahead and just took it out.  I'm still not certain there wouldn't be complications moving to a composition like this that aren't worth the trouble for something as straightforward as a lazy property, but we didn't actually need it.  And we can always add the composition in the future if we later decide to.